bazaar/src/lib.rs

```diff
@@ -0,0 +1,205 @@
+// TODO: Check if URL valid?
+// TODO: Add Article Upload / Removal
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        // initialisation
+      //  type Error = Error<T>;
+        fn deposit_event() = default;
+
+        /// Allow a user to create a shop
+        #[weight = 10_000]
+        pub fn upload_shop(origin, cid: CurrencyIdentifier, shop: ShopIdentifier) -> DispatchResult {
+            // Check that the extrinsic was signed and get the signer
+            let sender = ensure_signed(origin)?;
+
+            let mut owned_shops = ShopsOwned::<T>::get(cid, &sender);
+            let mut shops = ShopRegistry::get(cid);
+
+            // Verify that the specified shop has not already been created with fast search
+            ensure!(!ShopAffiliation::<T>::contains_key(cid, &shop), Error::<T>::ShopAlreadyCreated);   
+            
+             // TODO: Really necessary to do the binary search twice just to get index?
+            // Get the index of the last entry of the Shop vector          
+            match shops.binary_search(&shop) {
+                // If the search succeeds, the shop was already created
+                Ok(_) => Err(<Error<T>>::ShopAlreadyCreated.into()),
+                // If the search fails, the shop can be inserted into the owned list
+                Err(shop_registry_index) => {
+                    match owned_shops.binary_search(&shop) {
+                        Ok(_) => Err(<Error<T>>::ShopAlreadyCreated.into()), // should not be possible
+                        Err(onwed_shops_index) => {
+                            // Add the shop to the registries
+                            owned_shops.insert(onwed_shops_index, shop.clone());
+                            shops.insert(shop_registry_index, shop.clone());
+                            // Update blockchain        
+                            ShopsOwned::<T>::insert(cid, &sender, owned_shops);
+                            ShopAffiliation::<T>::insert(cid, &shop, &sender);
+                            ShopRegistry::insert(cid, shops);  
+                            // Emit an event that the shop was created
+                            Self::deposit_event(RawEvent::ShopCreated(sender, shop));
+                            Ok(())
+                        },
+                    }
+                },
+            }            
+        }
+
+        /// Allow a user to remove their shop.
+        #[weight = 10_000]
+        pub fn remove_shop(origin, cid:CurrencyIdentifier, shop: ShopIdentifier) -> DispatchResult {
+            // Check that the extrinsic was signed and get the signer.
+            let sender = ensure_signed(origin)?;
+
+            let mut owned_shops = ShopsOwned::<T>::get(cid, &sender);
+            let mut shops = ShopRegistry::get(cid);
+
+            // Verify that the specified shop is existing.
+            ensure!(ShopAffiliation::<T>::contains_key(cid, &shop), Error::<T>::NoSuchShop);
```

Comment:
why should the attempt to delete an inexisting shop cause an error? The result is the same